osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,237 @@
+                       << GetLastError();
+          continue;
+        }
+      }
+
+      std::vector<wchar_t> renderedContent(renderedBuffUsed);
+      renderedBuffSize = renderedBuffUsed;
+      if (!EvtRender(nullptr,
+                     events[i],
+                     EvtRenderEventXml,
+                     renderedBuffSize,
+                     renderedContent.data(),
+                     &renderedBuffUsed,
+                     &propCount)) {
+        LOG(WARNING) << "Failed to render windows event with "
+                     << GetLastError();
+        continue;
+      }
+
+      std::wstringstream xml_event;
+      xml_event << renderedContent.data();
+      parseWelXml(context, xml_event.str(), yield);
+      EvtClose(events[i]);
+    }
+
+    ret = EvtNext(
+        queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+  }
+  if (ERROR_NO_MORE_ITEMS != GetLastError()) {
+    // No need to close the handler after error; The query
+    // EvtClose will also close all the event handler
+    VLOG(1) << "EvtNext failed with error " << GetLastError();
+  }
+}
+
+void genXfilterFromConstraints(QueryContext& context, std::string& xfilter) {
+  std::vector<std::string> xfilterList;
+
+  auto eids = context.constraints["eventid"].getAll(EQUALS);
+  if (!eids.empty()) {
+    xfilterList.emplace_back(
+        "(EventID=" + osquery::join(eids, ") or (EventID=") + ")");
+  }
+
+  auto pids = context.constraints["pid"].getAll(EQUALS);
+  if (!pids.empty()) {
+    xfilterList.emplace_back(
+        "(Execution[@ProcessID=" +
+        osquery::join(pids, "]) or (Execution[@ProcessID=") + "])");
+  }
+
+  auto times = context.constraints["time_range"].getAll(EQUALS);
+  auto timestamps = context.constraints["timestamp"].getAll(EQUALS);
+  if (!times.empty()) {
+    auto datetime = *times.begin();
+    auto time_vec = osquery::split(datetime, ";");
+
+    if (time_vec.size() == 1) {
+      auto _start = time_vec.front();
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start + "']");
+    } else if (time_vec.size() == 2) {
+      auto _start = time_vec.front();
+      auto _end = time_vec.at(1);
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start +
+                               "' and @SystemTime&lt;='" + _end + "']");
+    }
+  } else if (!timestamps.empty()) {
+    auto time_diff = *timestamps.begin();
+    xfilterList.emplace_back(
+        "TimeCreated[timediff(@SystemTime) &lt;= " + time_diff + "]");
+  }
+
+  xfilter = xfilterList.empty()
+                ? "*"
+                : "*[System[" + osquery::join(xfilterList, " and ") + "]]";
+}
+
+void genWindowsEventLog(RowYield& yield, QueryContext& context) {
+  std::set<std::pair<std::string, std::string>> xpath_set;
+
+  // Check if the `xpath` constraint is available. The `xpath` has
+  // priority over `channel` and it will lookup and filter the events
+  // based on `xpath` if passed.
```

Comment:
I don't know if this should silently overwrite, vs error. Curious about the reasoning? And what the crowd thinks? I think I'm fine with most things, but I'd like to understand why

Reply:
I'm inclined to agree with @directionless. I think it should error if multiple conflicting options are passed.